Soapy SDRPlay Source fails with AGC=True

### What happened?

I changed the RF Options / AGC to `True` and the source doesn't start. The warning and error in log output do not show with AGC=False

Expected automatic gain to be enabled along with a new parameter to choose a level, AFAIK there are parameters to adjust the AGC.

### System Information

OS: Mac
GR Installation Method: homebrew

### GNU Radio Version

3.10 (maint-3.10)

### Specific Version

3.10.6.0

### Steps to Reproduce the Problem

1. Add Soapy SDRPlay Source
2. Edit Preferences; set Center Freq, and change AGC to `True`
3. Add Null Sink and connect
4. Execute

### Relevant log output

```shell
[INFO] devIdx: 0
[INFO] SerNo: B0002P0001
[INFO] hwVer: 1
[INFO] rspDuoMode: 0
[INFO] tuner: 1
[INFO] rspDuoSampleFreq: 0.000000
[WARNING] Not updating IFGR gain because AGC is enabled
[INFO] Using format CF32.
[ERROR] error in activateStream() - Init() failed: sdrplay_api_Fail



SoapySDRUtil -probe show following:

----------------------------------------------------
-- Device identification
----------------------------------------------------
  driver=SDRplay
  hardware=RSP1
  sdrplay_api_api_version=3.070000
  sdrplay_api_hw_version=1

----------------------------------------------------
-- Peripheral summary
----------------------------------------------------
  Channels: 1 Rx, 0 Tx
  Timestamps: NO
  Other Settings:
     * RF Gain Select - RF Gain Select
       [key=rfgain_sel, default=1, type=string, options=(0, 1, 2, 3)]
     * IQ Correction - IQ Correction Control
       [key=iqcorr_ctrl, default=true, type=bool]
     * AGC Setpoint - AGC Setpoint (dBfs)
       [key=agc_setpoint, default=-30, type=int, range=[-60, 0]]

----------------------------------------------------
-- RX Channel 0
----------------------------------------------------
  Full-duplex: NO
  Supports AGC: YES
  Stream formats: CS16, CF32
  Native format: CS16 [full-scale=32767]
  Antennas: RX
  Corrections: DC removal
  Full gain range: [0, 42] dB
    IFGR gain range: [20, 59] dB
    RFGR gain range: [0, 3] dB
  Full freq range: [0.01, 2000] MHz
    RF freq range: [0.01, 2000] MHz
    CORR freq range:  MHz
  Sample rates: 0.0625, 0.096, 0.125, 0.192, 0.25, ..., 6, 7, 8, 9, 10 MSps
  Filter bandwidths: 0.2, 0.3, 0.6, 1.536, 5, 6, 7, 8 MHz
```

Didn't work out of the box, default gain needed some tweaking for me
```
[WARNING] Not updating IFGR gain because AGC is enabled
source :error: Gain RFGR out of range: 0 <= gain <= 3
[INFO] Using format CF32.
[ERROR] error in activateStream() - Init() failed: sdrplay_api_Fail
```
I think usable range depends on the actual device so maybe there is no universally good default. The error message is little unhelpful as it doesn't correspond with any of the settings really.

--- a/gr-soapy/grc/soapy_sdrplay_source.block.yml
+++ b/gr-soapy/grc/soapy_sdrplay_source.block.yml
@@ -35,6 +35,12 @@ parameters:
     default: 'RX'
     hide: part
 
+  - id: bias
+    label: Bias Tee Power
+    dtype: bool
+    default: 'False'
+    hide: part
+
   - id: center_freq
     label: 'Center Freq (Hz)'
     category: RF Options
@@ -55,12 +61,26 @@ parameters:
     default: 'False'
     hide: part
 
-  - id: gain
-    label: 'RF Gain (-47dB - 0dB)'
+  - id: agc_setpoint
+    label: AGC Setpoint (-60dB - 0dB)'
     category: RF Options
-    dtype: real
-    default: '-20'
-    hide: ${'all' if agc else 'part'}
+    dtype: int
+    default: '-30'
+    hide: part
+
+  - id: gain
+    label: 'Gain (0dB - 39dB)'
+    category: RF Options
+    dtype: int
+    default: '20'
+    hide: part
+
+  - id: lna_state
+    label: 'LNA State (atten step)'
+    category: RF Options
+    dtype: int
+    default: '3'
+    hide: part
 
 inputs:
   - domain: message
@@ -74,29 +94,64 @@ outputs:
 
 templates:
   imports: from gnuradio import soapy
-  make: |
+
+  make: |-
       None
+      ## Negative value for agc_setpoint is placed in parens by GRC.
+      ## This code turns it into an int and bounds checks
+      _agc_setpoint = int(${agc_setpoint})
+      _agc_setpoint = max(min(_agc_setpoint, -20), -70)
+
       dev = 'driver=sdrplay'
       stream_args = ''
       tune_args = ['']
       settings = ['']
+
+      ## Intercept the agc callback and restore the gain when agc is
+      ## disabled. The driver does not do this. LNA state is preserved.
+      def _set_${id}_gain_mode(channel, agc):
+          self.${id}.set_gain_mode(channel, agc)
+          if not agc:
+              self.set_${id}_gain(channel, self.gain)
+      self.set_${id}_gain_mode = _set_${id}_gain_mode
+
+      ## Intercept the gain callback. The soapysdrplay driver prints an
+      ## error message every time gain is adjusted with agc on. Defer
+      ## setting IFGR until agc is turned off. Bounds check.
+      def _set_${id}_gain(channel, gain):
+          if not self.agc:
+              self.${id}.set_gain(channel, 'IFGR', min(max(59 - gain, 20), 59))
+      self.set_${id}_gain = _set_${id}_gain
+
+      ## Valid values for LNA state (RFGR here) vary by model, so runtime
+      ## changes may cause errors.
+      def _set_${id}_lna_state(channel, lna_state):
+              self.${id}.set_gain(channel, 'RFGR', min(max(lna_state, 0), 9))
+      self.set_${id}_lna_state = _set_${id}_lna_state
 
       self.${id} = soapy.source(dev, "${type}", 1, ${dev_args},
                                 stream_args, tune_args, settings)
       self.${id}.set_sample_rate(0, ${samp_rate})
       self.${id}.set_bandwidth(0, ${bandwidth})
       self.${id}.set_antenna(0, ${antenna})
-      self.${id}.set_gain_mode(0, ${agc})
       self.${id}.set_frequency(0, ${center_freq})
       self.${id}.set_frequency_correction(0, ${freq_correction})
-      self.${id}.set_gain(0, min(max(-${gain}, 0.0), 47.0))
+      self.${id}.write_setting('biasT_ctrl', ${bias})
+      self.${id}.write_setting('agc_setpoint', ${agc_setpoint})
+      self.set_${id}_gain_mode(0, ${agc})
+      self.set_${id}_gain(0, ${gain})
+      self.set_${id}_lna_state(0, ${lna_state})
+
   callbacks:
     - set_sample_rate(0, ${samp_rate})
     - set_bandwidth(0, ${bandwidth})
     - set_antenna(0, ${antenna})
-    - set_gain_mode(0, ${agc})
     - set_frequency(0, ${center_freq})
     - set_frequency_correction(0, ${freq_correction})
-    - set_gain(0, min(max(-${gain}, 0.0), 47.0))
+    - write_setting('biasT_ctrl', ${bias})
+    - write_setting('agc_setpoint', ${agc_setpoint})
+    - self.set_${id}_gain_mode(0, ${agc})
+    - self.set_${id}_gain(0, ${gain})
+    - self.set_${id}_lna_state(0, ${lna_state})
 
 file_format: 1

